File: app/src/main/java/io/geph/android/tun2socks/TunnelManager.java
package io.geph.android.tun2socks;

import android.annotation.TargetApi;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.VpnService;
import android.os.Build;
import android.os.ParcelFileDescriptor;
import android.support.v4.app.NotificationCompat;
import android.util.Log;

import java.io.BufferedInputStream;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicBoolean;

import io.geph.android.MainActivity;
import io.geph.android.R;

public class TunnelManager  {

    public static final int NOTIFICATION_ID = 7839214;

    public static final String SOCKS_SERVER_ADDRESS_BASE = "socksServerAddress";
    public static final String SOCKS_SERVER_PORT_EXTRA = "socksServerPort";
    public static final String DNS_SERVER_PORT_EXTRA = "dnsServerPort";
    public static final String USERNAME = "username";
    public static final String PASSWORD = "password";
    public static final String EXIT_NAME = "exitName";
    public static final String LISTEN_ALL = "listenAll";
    public static final String FORCE_BRIDGES = "forceBridges";
    public static final String BYPASS_CHINESE = "bypassChinese";

    private static final String LOG_TAG = "TunnelManager";
    private static final String CACHE_DIR_NAME = "geph";
    private static final String DAEMON_IN_NATIVELIB_DIR = "libgeph.so";
    private TunnelVpnService m_parentService = null;
    private CountDownLatch m_tunnelThreadStopSignal;
    private Thread m_tunnelThread;
    private AtomicBoolean m_isStopping;
    private String mSocksServerAddressBase;
    private String mSocksServerAddress;
    private String mSocksServerPort;
    private String mDnsServerPort;
    private String mDnsResolverAddress;
    private String mUsername;
    private String mPassword;
    private String mExitName;
    private Boolean mListenAll;
    private Boolean mForceBridges;
    private Boolean mBypassChinese;
    private Process mDaemonProc;
    private AtomicBoolean m_isReconnecting;

    public TunnelManager(TunnelVpnService parentService) {
        m_parentService = parentService;
        m_isStopping = new AtomicBoolean(false);
        m_isReconnecting = new AtomicBoolean(false);
    }

    // Implementation of android.app.Service.onStartCommand
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent == null) {
            Log.i(LOG_TAG, "labooyah");
            return 0;
        }
        Log.i(LOG_TAG, "onStartCommand");

        mSocksServerAddressBase = intent.getStringExtra(SOCKS_SERVER_ADDRESS_BASE);
        mSocksServerPort = intent.getStringExtra(SOCKS_SERVER_PORT_EXTRA);
        mSocksServerAddress = mSocksServerAddressBase + ":" + mSocksServerPort;
        mDnsServerPort = intent.getStringExtra(DNS_SERVER_PORT_EXTRA);
        mDnsResolverAddress = mSocksServerAddressBase + ":" + mDnsServerPort;
        Log.i(LOG_TAG, "onStartCommand parsed some stuff");
        mUsername = intent.getStringExtra(USERNAME);
        mPassword = intent.getStringExtra(PASSWORD);
        mExitName = intent.getStringExtra(EXIT_NAME);
        mForceBridges = intent.getBooleanExtra(FORCE_BRIDGES, false);
        mListenAll = intent.getBooleanExtra(LISTEN_ALL, false);
        mBypassChinese = intent.getBooleanExtra(BYPASS_CHINESE, false);
        Log.i(LOG_TAG, "onStartCommand parsed intent");

        if (setupAndRunDaemon() == null) {
            Log.e(LOG_TAG, "Failed to start the socks proxy daemon.");
            m_parentService.broadcastVpnStart(false /* success */);
            return 0;
        }

        if (mSocksServerAddress == null) {
            Log.e(LOG_TAG, "Failed to receive the socks server address.");
            m_parentService.broadcastVpnStart(false /* success */);
            return 0;
        }

        final Context ctx = getContext();

        Intent notificationIntent = new Intent(ctx, MainActivity.class);
        PendingIntent pendingIntent = PendingIntent.getActivity(ctx, 0, notificationIntent, 0);

        Bitmap largeIcon = BitmapFactory.decodeResource(ctx.getResources(), R.mipmap.ic_launcher);
        String channelId = createNotificationChannel();
        NotificationCompat.Builder builder = new NotificationCompat.Builder(ctx, createNotificationChannel())
                .setSmallIcon(R.drawable.ic_stat_notification_icon)
                .setLargeIcon(largeIcon)
                .setWhen(System.currentTimeMillis())
                .setContentTitle(ctx.getText(R.string.notification_label))
                .setAutoCancel(true)
                .setContentIntent(pendingIntent);

        Notification notification = builder.build();

        // starting this service on foreground to avoid accidental GC by Android system
        getVpnService().startForeground(NOTIFICATION_ID, notification);

        return Service.START_STICKY;
    }

    private String createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            String channelId = "geph_service";
            String channelName = "Geph background service";
            NotificationChannel chan = new NotificationChannel(channelId,
                    channelName, NotificationManager.IMPORTANCE_NONE);
            chan.setDescription("Geph background service");
            NotificationManager notificationManager = getContext().getSystemService(NotificationManager.class);
            assert notificationManager != null;
            notificationManager.createNotificationChannel(chan);
            return channelId;
        }
        return "";
    }

    private Process setupAndRunDaemon() {
        // Run Geph-Go with uProxy style fail-safe
        mDaemonProc = runDaemon();
        if (mDaemonProc == null) {
            return mDaemonProc;
        }

        Thread daemonLogs = new Thread(new Runnable() {
            @Override
            public void run() {
                String tag = "Geph";
                InputStream err = mDaemonProc.getErrorStream();
                Scanner scanner = new Scanner(err);
                while (scanner.hasNextLine()) {
                    String line = scanner.nextLine();
                    Log.d(tag, line);
                }
                Log.d(tag, "stopping log stuff because the process died");
                getVpnService().stopForeground(true);
                System.exit(0);
            }
        });
        daemonLogs.start();

        Thread daemonStdout = new Thread(new Runnable() {
            @Override
            public void run() {
                final String tag = "GephStdio";
                ParcelFileDescriptor tunFd = null;
                FileInputStream input = null;
                FileOutputStream output = null;
                try {
                    try (InputStream raw_stream = mDaemonProc.getInputStream()) {
                        BufferedInputStream stream = new BufferedInputStream(raw_stream);
                        Thread daemonStdin = null;
                        while (true) {
                            int kind = stream.read();
                            int bodylen1 = stream.read();
                            int bodylen2 = stream.read();
                            int bodylen = bodylen1 + bodylen2*256;
                            byte[] body = new byte[bodylen];
                            int n = 0;
                            while (n < body.length) {
                                n += stream.read(body, n, body.length - n);
                            }
                            if (kind == 1) {
                                String bodyString = new String(body, StandardCharsets.UTF_8);
                                Log.d(tag, "init IP address of " + bodyString);
                                String[] splitted = bodyString.split("/");
                                InetAddress addr = InetAddress.getByName(splitted[0]);
                                if (tunFd != null) {
                                    input.close();
                                    output.close();
                                    tunFd.close();
                                    daemonStdin.interrupt();
                                    daemonStdin.wait();
                                }
                                tunFd = m_parentService.newBuilder().addAddress(addr, 10).
                                        addRoute("0.0.0.0", 0).
                                        addDnsServer("9.9.9.9").
                                        addDisallowedApplication(getContext().getPackageName())
                                        .setBlocking(true)
                                        .setMtu(1280)
                                        .establish();
                                assert tunFd != null;
                                input = new FileInputStream(tunFd.getFileDescriptor());
                                output = new FileOutputStream(tunFd.getFileDescriptor());
                                final FileInputStream finalInput = input;
                                daemonStdin = new Thread(new Runnable() {
                                    @Override
                                    public void run() {
                                        byte[] body = new byte[2048];
                                        try {
                                            try (OutputStream stream = mDaemonProc.getOutputStream()) {
                                                while (true) {
                                                    try {
                                                        int n = finalInput.read(body);
                                                        stream.write(0);
                                                        stream.write(n % 256);
                                                        stream.write(n / 256);
                                                        stream.write(body, 0, n);
                                                        stream.flush();
                                                    } catch (IOException e) {
                                                        e.printStackTrace();
                                                    }
                                                }
                                            }
                                        } catch (IOException e) {
                                            e.printStackTrace();
                                        }
                                    }
                                });
                                daemonStdin.start();
                            } else {
                                if (output != null) {
                                    try {
                                        output.write(body);
                                    } catch (Exception e) {
                                        e.printStackTrace();
                                    }
                                }
                            }
                        }
                    }
                } catch (IOException | PackageManager.NameNotFoundException | InterruptedException e) {
                    e.printStackTrace();
                }
                finally {
                    if (tunFd != null) {
                        try {
                            tunFd.close();
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }
                }
            }
        });
        daemonStdout.start();

        return mDaemonProc;
    }

    private Process runDaemon() {
        final Context ctx = getContext();
        final String daemonBinaryPath =
                ctx.getApplicationInfo().nativeLibraryDir + "/" + DAEMON_IN_NATIVELIB_DIR;
        final String dbPath = ctx.getApplicationInfo().dataDir + "/geph4-credentials.db";

        try {
            List<String> commands = new ArrayList<>();
            commands.add(daemonBinaryPath);
            commands.add("connect");
            commands.add("--username");
            commands.add(mUsername);
            commands.add("--password");
            commands.add(mPassword);
            commands.add("--exit-server");
            commands.add(mExitName);
            commands.add("--credential-cache");
            commands.add(dbPath);
            commands.add("--dns-listen");
            commands.add("127.0.0.1:15353");
            commands.add("--stdio-vpn");
//            commands.add("-fakeDNS=true");
//            commands.add("-dnsAddr=127.0.0.1:49983");
            if (mListenAll) {
                commands.add("--socks5-listen");
                commands.add("0.0.0.0:9909");
            }
            if (mForceBridges) {
                commands.add("--use-bridges");
            }
            if (mBypassChinese) {
                commands.add("--exclude-prc");
            }
            Log.i(LOG_TAG, commands.toString());
            ProcessBuilder pb = new ProcessBuilder(commands);

            return pb.start();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public void terminateSocksProxyDaemon() {
        if (mDaemonProc != null) {
            mDaemonProc.destroy();
            mDaemonProc = null;
        }
    }

    public void restartSocksProxyDaemon() {
        terminateSocksProxyDaemon();
        setupAndRunDaemon();
    }

    // Implementation of android.app.Service.onDestroy
    public void onDestroy() {
        terminateSocksProxyDaemon();

        if (m_tunnelThread == null) {
            return;
        }

        // signalStopService should have been called, but in case is was not, call here.
        // If signalStopService was not already called, the join may block the calling
        // thread for some time.
        signalStopService();

        try {
            m_tunnelThread.join();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        m_tunnelThreadStopSignal = null;
        m_tunnelThread = null;

        // stop the foreground service
        getVpnService().stopForeground(true);
    }

    // Signals the runTunnel thread to stopUi. The thread will self-stopUi the service.
    // This is the preferred method for stopping the tunnel service:
    // 1. VpnService doesn't respond to stopService calls
    // 2. The UI will not block while waiting for stopService to return
    public void signalStopService() {
        if (m_tunnelThreadStopSignal != null) {
            m_tunnelThreadStopSignal.countDown();
        }
    }

    // Stops the tunnel thread and restarts it with |socksServerAddress|.
    public void restartTunnel(final String socksServerAddress, String socksServerPort, String dnsServerPort) {
        Log.i(LOG_TAG, "Restarting tunnel.");
        if (socksServerAddress == null ||
                socksServerAddress.equals(mSocksServerAddress)) {
            // Don't reconnect if the socks server address hasn't changed.
            m_parentService.broadcastVpnStart(true /* success */);
            return;
        }
        mSocksServerAddress = socksServerAddress;
        mSocksServerPort = socksServerPort;
        mDnsServerPort = dnsServerPort;

        m_isReconnecting.set(true);

        // Signaling stopUi to the tunnel thread with the reconnect flag set causes
        // the thread to stopUi the tunnel (but not the VPN or the service) and send
        // the new SOCKS server address to the DNS resolver before exiting itself.
        // When the DNS broadcasts its local address, the tunnel will restart.
        signalStopService();
    }


    private void runTunnel(String socksServerAddress, String dnsResolverAddress) {
        m_isStopping.set(false);

        try {
            Log.i(LOG_TAG, "VPN service running");
            m_parentService.broadcastVpnStart(true /* success */);

            try {
                m_tunnelThreadStopSignal.await();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }

            m_isStopping.set(true);

        } finally {
                m_parentService.stopForeground(true);
                m_parentService.stopSelf();
        }
    }

    //----------------------------------------------------------------------------
    // Tunnel.HostService
    //----------------------------------------------------------------------------


    public Context getContext() {
        return m_parentService;
    }

    public VpnService getVpnService() {
        return ((TunnelVpnService) m_parentService);
    }

    public VpnService.Builder newVpnServiceBuilder() {
        return ((TunnelVpnService) m_parentService).newBuilder();
    }

    public void onDiagnosticMessage(String message) {
        Log.d(LOG_TAG, message);
    }

    public void onTunnelConnected() {
        Log.i(LOG_TAG, "Tunnel connected.");
    }

    @TargetApi(Build.VERSION_CODES.M)
    public void onVpnEstablished() {
        Log.i(LOG_TAG, "VPN established.");
    }
}

File: app/src/main/java/io/geph/android/tun2socks/TunnelManager.kt
package io.geph.android.tun2socks

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.PendingIntent.FLAG_IMMUTABLE
import android.app.Service
import android.content.pm.ServiceInfo
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.os.Build
import android.os.ParcelFileDescriptor
import android.os.Process
import android.system.OsConstants.F_SETFD
import android.util.Log
import androidx.core.app.NotificationCompat
import com.frybits.harmony.getHarmonySharedPreferences
import com.sun.jna.Library
import com.sun.jna.Native
import io.geph.android.DaemonArgs
import io.geph.android.GephDaemon
import io.geph.android.MainActivity
import io.geph.android.R
import io.geph.android.ReviewPromptState
import kotlinx.serialization.json.*
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import kotlin.concurrent.thread


interface LibC : Library {
    fun fcntl(fd: Int, cmd: Int, args: Int): Int
    fun dup2(oldFd: Int, newFd: Int): Int
}

class TunnelManager(parentService: TunnelVpnService?) {
    private var parentService: TunnelVpnService? = parentService
    private var tunFd: ParcelFileDescriptor? = null
    private var gephDaemon: GephDaemon? = null
    // How the engine reaches the tun: 0 = inherits the app's fd 0 (API 26+,
    // dup2'd once at first start and never closed), -1 = legacy stdio pump.
    private var engineFd = -1
    private var reviewSessionRecorded = false

    fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent == null) {
            Log.i(LOG_TAG, "Intent is null")
            return 0
        }
        Log.i(LOG_TAG, "onStartCommand")

        // Setup and run VPN service with daemon
        setupAndRunVpnService()
        
        // Set up notification
        val ctx = requireContext()
        val notificationIntent = Intent(ctx, MainActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(ctx, 0, notificationIntent, FLAG_IMMUTABLE)
        val largeIcon = BitmapFactory.decodeResource(ctx.resources, R.mipmap.ic_launcher)
        val channelId = createNotificationChannel()
        val builder = NotificationCompat.Builder(ctx, channelId)
                .setSmallIcon(R.drawable.ic_stat_notification_icon)
                .setLargeIcon(largeIcon)
                .setWhen(System.currentTimeMillis())
                .setContentTitle(ctx.getText(R.string.notification_label))
                .setAutoCancel(true)
                .setContentIntent(pendingIntent)
        val notification = builder.build()
        
        // Start foreground as special-use on API 34+ (VPNs require continuous FGS)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) { // API 34+
            requireVpnService().startForeground(
                NOTIFICATION_ID,
                notification,
                ServiceInfo.FOREGROUND_SERVICE_TYPE_SPECIAL_USE
            )
        } else {
            requireVpnService().startForeground(NOTIFICATION_ID, notification)
        }
        return Service.START_STICKY
    }

    private fun createNotificationChannel(): String {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channelId = "geph_service"
            val channelName = "Geph background service"
            val chan = NotificationChannel(channelId,
                    channelName, NotificationManager.IMPORTANCE_NONE)
            chan.description = "Geph background service"
            val notificationManager = requireContext().getSystemService(NotificationManager::class.java)
            notificationManager.createNotificationChannel(chan)
            return channelId
        }
        return ""
    }

    private fun setupAndRunVpnService() {
        Log.e("SETUP", "Setting up VPN service and daemon")
        
        // Get DaemonArgs from shared preferences
        val prefs = requireContext().getHarmonySharedPreferences("daemon")
        val daemonArgsJson = prefs.getString(DAEMON_ARGS, null)
        
        if (daemonArgsJson == null) {
            Log.e(LOG_TAG, "No daemon arguments found in preferences")
            parentService?.broadcastVpnStart(false)
            return
        }
        
        // Parse DaemonArgs from JSON
        val daemonArgs = try {
            val json = Json { ignoreUnknownKeys = true }
            json.decodeFromString(DaemonArgs.serializer(), daemonArgsJson)
        } catch (e: Exception) {
            Log.e(LOG_TAG, "Failed to parse daemon arguments: ${e.message}")
            parentService?.broadcastVpnStart(false)
            return
        }
        
        // Create VPN interface
        var vpnInterface: ParcelFileDescriptor? = null
        while (vpnInterface == null) {
            Log.d("SETUP", "Attempting to create VPN interface")
            val builder = requireParentService().newBuilder()
                .addAddress("100.64.89.64", 10)
                .addRoute("0.0.0.0", 0)
                .addDnsServer("100.64.89.1")
                .addDisallowedApplication(requireContext().packageName)
            
            // Add excluded apps from the app whitelist
            try {
                // Only process app whitelist if it's not empty
                for (packageName in daemonArgs.appWhitelist) {
                    try {
                        builder.addDisallowedApplication(packageName)
                    } catch (e: Exception) {
                        Log.w(LOG_TAG, "Failed to add app to exclusion list: $packageName")
                    }
                }
            } catch (e: Exception) {
                Log.e(LOG_TAG, "Error setting up app exclusions: ${e.message}")
            }
            
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                builder.setMetered(false)
            }
            
            vpnInterface = builder.setBlocking(true)
                .setMtu(16384)
                .establish()
        }
        
        tunFd = vpnInterface
        startGephDaemon(vpnInterface, daemonArgs)
    }

    private fun startGephDaemon(vpnInterface: ParcelFileDescriptor, daemonArgs: DaemonArgs) {
        engineFd = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val fd = vpnInterface.detachFd()
            val libc = Native.load(LibC::class.java)
            libc.fcntl(fd, F_SETFD, 0)
            libc.dup2(fd, 0)
            0
        } else {
            -1 // Will handle stdio-based approach for older versions
        }

        spawnEngine(daemonArgs)

        if (!reviewSessionRecorded) {
            ReviewPromptState.recordVpnSessionStarted(requireContext())
            reviewSessionRecorded = true
        }

        // Broadcast successful VPN start
        parentService?.broadcastVpnStart(true)
    }

    /**
     * Spawn a geph5-client engine against the already-configured tun (the
     * app's fd 0 on API 26+, the stdio pump otherwise). Used both for the
     * first start and for in-place restarts.
     */
    private fun spawnEngine(daemonArgs: DaemonArgs) {
        // Create a config from the DaemonArgs
        val config = daemonArgs.toConfig(requireContext()).jsonObject

        val controlSockPath = vpnControlSockPath(requireContext())

        // Add VPN-specific configurations to the config
        val vpnEnabledConfig = buildJsonObject {
            // Copy all elements from the original config
            for ((key, value) in config) {
                put(key, value)
            }

            // Add VPN fd configuration if available
            if (engineFd >= 0) {
                put("vpn_fd", engineFd)
            }

            // Serve the control RPC over a Unix-domain socket in the app's
            // private files dir, instead of a localhost TCP port that other
            // apps on the device could connect to.
            put("control_listen_unix", controlSockPath)
        }

        // Create and start the daemon
        val daemon = GephDaemon(requireContext(), vpnEnabledConfig)
        gephDaemon = daemon

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            // For older Android versions, manually handle tunnel I/O
            startLegacyIo(tunFd!!, daemon)
        }

        // Monitor the engine for crashes. Only tear the service down if the
        // engine that exited is still the current one — an engine we replaced
        // in restartEngine() exits deliberately.
        thread {
            try {
                val exitCode = daemon.waitForExit()
                if (gephDaemon === daemon) {
                    Log.e(LOG_TAG, "Daemon process exited with code: $exitCode")
                    signalStopService()
                } else {
                    Log.i(LOG_TAG, "old engine exited with code $exitCode after restart")
                }
            } catch (e: Exception) {
                Log.e(LOG_TAG, "Error monitoring daemon process: ${e.message}")
            }
        }
    }

    /**
     * Swap the engine under the live tun: kill the current geph5-client and
     * spawn a fresh one from the (re-read) persisted DaemonArgs. The
     * VpnService and tun stay up the whole time, so traffic blackholes rather
     * than leaks during the gap. Used to apply a new exit while connected.
     */
    @Synchronized
    fun restartEngine() {
        val old = gephDaemon ?: run {
            Log.w(LOG_TAG, "restartEngine with no running engine; ignoring")
            return
        }

        val prefs = requireContext().getHarmonySharedPreferences("daemon")
        val daemonArgsJson = prefs.getString(DAEMON_ARGS, null)
        if (daemonArgsJson == null) {
            Log.e(LOG_TAG, "restartEngine: no daemon arguments in preferences")
            return
        }
        val daemonArgs = try {
            val json = Json { ignoreUnknownKeys = true }
            json.decodeFromString(DaemonArgs.serializer(), daemonArgsJson)
        } catch (e: Exception) {
            Log.e(LOG_TAG, "restartEngine: failed to parse daemon arguments: ${e.message}")
            return
        }

        // Detach first so the old engine's watchdog sees it is no longer
        // current and does not tear the service down.
        gephDaemon = null
        old.stopDaemon()
        spawnEngine(daemonArgs)
    }

    private fun startLegacyIo(vpnInterface: ParcelFileDescriptor, daemon: GephDaemon) {
        // The pumps capture the engine instance they serve (not the field), so
        // threads belonging to a replaced engine die with it instead of
        // fighting the new engine's pumps. The streams are never closed here:
        // closing them would close the shared tun fd, whose lifetime belongs
        // to tunFd (closed in terminateDaemon).
        // download
        Log.e(LOG_TAG, "VPN I/O SET UP")
        Thread {
            val body = ByteArray(40000)
            val writer = FileOutputStream(vpnInterface.fileDescriptor)
            try {
                while (!Thread.currentThread().isInterrupted && daemon.isAlive) {
                    val n = daemon.downloadVpn(body)
                    if (n <= 0) {
                        break
                    }
                    writer.write(body, 0, n)
                }
            } catch (e: Exception) {
                Log.d(LOG_TAG, "VPN download pump exited: ${e.message}")
            }
        }.start()
        // upload
        Thread {
            Log.e(LOG_TAG, "VPN I/O UP STARTED")
            val body = ByteArray(40000)
            val reader = FileInputStream(vpnInterface.fileDescriptor)
            try {
                while (!Thread.currentThread().isInterrupted && daemon.isAlive) {
                    val n = reader.read(body)
                    if (n <= 0) {
                        break
                    }
                    daemon.uploadVpn(body, n)
                }
            } catch (e: Exception) {
                Log.d(LOG_TAG, "VPN upload pump exited: ${e.message}")
            }
        }.start()
    }

    fun terminateDaemon() {
        parentService?.let { ReviewPromptState.recordVpnSessionEnded(it) }
        reviewSessionRecorded = false
        gephDaemon?.stopDaemon()
        gephDaemon = null
        
        // Close the VPN interface
        tunFd?.close()
        tunFd = null
    }

    fun onDestroy() {
        terminateDaemon()
        stopForegroundCompat()
    }

    fun signalStopService() {
        terminateDaemon()
        parentService?.broadcastVpnDisconnect()
        stopForegroundCompat()
        parentService?.stopSelf()
        Process.killProcess(Process.myPid())
    }

    companion object {
        const val NOTIFICATION_ID = 7839214

        // The key for storing DaemonArgs in SharedPreferences
        const val DAEMON_ARGS = "daemonArgs"

        private const val VPN_CONTROL_SOCK_NAME = "control.sock"
        private const val FALLBACK_CONTROL_SOCK_NAME = "control-fallback.sock"

        /** Filesystem path of the VPN daemon's control RPC Unix socket. */
        @JvmStatic
        fun vpnControlSockPath(context: Context): String =
            File(context.filesDir, VPN_CONTROL_SOCK_NAME).absolutePath

        /** Filesystem path of the fallback (non-VPN) daemon's control RPC Unix socket. */
        @JvmStatic
        fun fallbackControlSockPath(context: Context): String =
            File(context.filesDir, FALLBACK_CONTROL_SOCK_NAME).absolutePath

        private const val LOG_TAG = "TunnelManager"
    }

    private fun requireContext(): Context = checkNotNull(parentService) { "VPN service unavailable" }

    private fun requireParentService(): TunnelVpnService =
        checkNotNull(parentService) { "VPN service unavailable" }

    private fun requireVpnService(): TunnelVpnService =
        checkNotNull(parentService) { "VPN service unavailable" }

    private fun stopForegroundCompat() {
        val service = parentService ?: return
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            service.stopForeground(Service.STOP_FOREGROUND_REMOVE)
        } else {
            @Suppress("DEPRECATION")
            service.stopForeground(true)
        }
    }
}
